### ISSUE TITLE
AWS.jl ignores "location" from optional parameters in API request structures

### ISSUE THREAD

**Issue by omus** [JuliaCloud/AWS.jl#536]
When working with the [S3 GetObject API call](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#API_GetObject_RequestSyntax) I wanted to use the "Range" parameter which must be specified as a header. In attempting to specify this parameter in `S3.get_object` I discovered the request being made added the parameter in the HTTP query string and not the header. The API JSON files we use to generate the high-level code do specify the location where the parameter must be defined:

```json
    "GetObjectRequest": {
      "type": "structure",
      "required": [
        "Bucket",
        "Key"
      ],
      "members": {
        "Key": {
          "shape": "ObjectKey",
          "documentation": "<p>Key of the object to get.</p>",
          "location": "uri",
          "locationName": "Key"
        },
        "Range": {
          "shape": "Range",
          "documentation": "<p>Downloads the specified range bytes of an object. For more information about the HTTP Range header, see <a href=\"https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35\">https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35</a>.</p> <note> <p>Amazon S3 doesn't support retrieving multiple ranges of data per <code>GET</code> request.</p> </note>",
          "location": "header",
          "locationName": "Range"
        },
     }
```

– https://github.com/aws/aws-sdk-js/blob/d05708ee0b52974be66b185410fde1900ee9a1c0/apis/s3-2006-03-01.normal.json#L5301-L5312

So this issue appears to be an oversight on our part.

**Comment by omus** [comment]
The two solutions I see right now for tackling this is:

- Add a `headers` argument to the low-level functions so that headers can be passed in separately from URI query args
- Make low-level functions "location" aware. Users of low-level functions would continue to pass in just `args` and the internals would then assign the parameters to the correct location

We could actually do both of these things at the same time but this would involve some restructuring

**Comment by mattBrzezinski** [comment]
Right.. originally when looking at these I was unable to find examples looking through of optional parameters required in specific positions. I was only able to find required ones in the `uri` or `headers`.

You can work around this currently by doing the low-level requests and in the `args::Dict()` specify `headers => Dict()` when making your request.

I believe that should make the request work as you intend.

**Comment by omus** [comment]
> You can work around this currently by doing the low-level requests and in the `args::Dict()` specify `headers => Dict()` when making your request.

Definitely a work around. Ideally high-level end users wouldn't need to be aware of the location the API parameter needs to used in. Especially as our documentation for these functions doesn't mention these details.

**Comment by mattBrzezinski** [comment]
> > You can work around this currently by doing the low-level requests and in the `args::Dict()` specify `headers => Dict()` when making your request.
> 
> Definitely a work around. Ideally high-level end users wouldn't need to be aware of the location the API parameter needs to used in. Especially as our documentation for these functions doesn't mention these details.

I agree. This feels like a change that's somewhat inline with modifying how optional args are passed along. We could do something similar with how required arguments are but just include a check if they are present and place them in their proper position.

Example, [S3.upload_part_copy()](https://github.com/JuliaCloud/AWS.jl/blob/6080e9d342d2e38eb92122aaedd2d1cfb813fead/src/services/s3.jl#L6303)
